src/backend/distributed/deparser/citus_ruleutils.c

```diff
@@ -1353,3 +1354,139 @@ RoleSpecString(RoleSpec *spec, bool withQuoteIdentifier)
 		}
 	}
 }
+
+
+/*
+ * ExpandMergedSubscriptingRefEntries takes a list of target entries and expands
+ * each one that references a SubscriptingRef node that indicates multiple (field)
+ * updates on the same attribute, which is applicable for array/json types atm.
+ */
+List *
+ExpandMergedSubscriptingRefEntries(List *targetEntryList)
+{
+	List *newTargetEntryList = NIL;
+
+	TargetEntry *targetEntry = NULL;
+	foreach_ptr(targetEntry, targetEntryList)
+	{
+		List *expandedTargetEntries = NIL;
+
+		Expr *expr = targetEntry->expr;
+		while (expr)
+		{
+			SubscriptingRef *subsRef = TargetListEntryExprFindSubscriptingRef(expr);
+			if (!subsRef)
+			{
+				break;
+			}
+
+			/*
+			 * Remove refexpr from the SubscriptingRef that we are about to
+			 * wrap in a new TargetEntry and save it for the next one.
+			 */
+			Expr *refexpr = subsRef->refexpr;
+			subsRef->refexpr = NULL;
+
+			/*
+			 * Wrap the Expr that holds SubscriptingRef (directly or indirectly)
+			 * in a new TargetEntry; note that it doesn't have a refexpr anymore.
+			 */
+			TargetEntry *newTargetEntry = copyObject(targetEntry);
+			newTargetEntry->expr = expr;
+			expandedTargetEntries = lappend(expandedTargetEntries, newTargetEntry);
+
+			/* now inspect the refexpr that SubscriptingRef at hand were holding */
+			expr = refexpr;
+		}
+
+		if (expandedTargetEntries == NIL)
+		{
+			/* return original entry since it doesn't hold a SubscriptingRef node */
+			newTargetEntryList = lappend(newTargetEntryList, targetEntry);
+		}
+		else
+		{
+			/*
+			 * Need to concat expanded target list entries in reverse order
+			 * to preserve ordering of the original target entry list.
+			 */
+			newTargetEntryList = list_concat(newTargetEntryList,
+											 list_reverse(expandedTargetEntries));
+		}
+	}
+
+	return newTargetEntryList;
+}
+
+
+/*
+ * TargetListEntryExprFindSubscriptingRef searches given Expr --assuming that
+ * it is part of a target list entry-- to see if it directly (i.e.: itself)
+ * or indirectly (e.g.: behind some level of coercions) holds a SubscriptingRef
+ * node.
+ *
+ * Returns the original SubscriptingRef node on success or NULL otherwise.
+ *
+ * Note that it wouldn't add much value to use expression_tree_walker here
+ * since we are only interested in a subset of the fields of a few certain
+ * node types.
+ */
+static SubscriptingRef *
+TargetListEntryExprFindSubscriptingRef(Expr *expr)
```

Comment:
Maybe get rid of this whole function since we cannot encounter with a `FieldStore`/`CoerceToDomain` node here atm, but I don't want to lose the findings noted here, let me think of a better way to document that.

Also, it's good to have those error checks as extra seat belts here.

Reply:
Or maybe we can keep this function as is and add a few tests using `deparse_shard_query` to show that it's indeed possible to see such nodes here but it cannot be the case due to the checks made in distributed planner beforehand ?

Reply:
> Maybe get rid of this whole function since we cannot encounter with a `FieldStore`/`CoerceToDomain` node here atm, but I don't want to lose the findings noted here, let me think of a better way to document that.
> 
> Also, it's good to have those error checks as extra seat belts.

Theoretically, we can still encounter with a `CoerceToDomain` node, now looking into that to see how to test this.

Reply:
guess I found how to test this :) ;

```sql
CREATE DOMAIN single_int_dom AS int[] CHECK ((VALUE[0]) > 0);

SELECT run_command_on_workers(
$$
    CREATE DOMAIN single_int_dom AS int[] CHECK ((VALUE[0]) > 0);
$$);

CREATE TABLE single_int_dom_indirection_test (id int, single_int_dom_col single_int_dom);
SELECT create_distributed_table('single_int_dom_indirection_test', 'id');

SELECT deparse_shard_query_test($$
    UPDATE single_int_dom_indirection_test
    SET single_int_dom_col[0] = 5;
$$);
```